awx/main/analytics/collectors.py

```diff
@@ -0,0 +1,168 @@
+            num_teams=Count('teams', distinct=True)):  # use .only()
+        counts[org.id] = {'name': org.name,
+                          'users': org.num_users,
+                          'teams': org.num_teams
+                          }
+    return counts
+    
+    
+@register('cred_type_counts')
+def cred_type_counts(since):
+    counts = {}
+    for cred_type in models.CredentialType.objects.annotate(
+            num_credentials=Count('credentials', distinct=True)):  
+        counts[cred_type.id] = {'name': cred_type.name,
+                                'credential_count': cred_type.num_credentials
+                                }
+    return counts
+    
+    
+@register('inventory_counts')
+def inventory_counts(since):
+    counts = {}
+    from django.db.models import Count
+    for inv in models.Inventory.objects.annotate(
+            num_sources=Count('inventory_sources', distinct=True), 
+            num_hosts=Count('hosts', distinct=True)).only('id', 'name', 'kind'):
+        counts[inv.id] = {'name': inv.name,
+                          'kind': inv.kind,
+                          'hosts': inv.num_hosts,
+                          'sources': inv.num_sources
+                          }
+    return counts
+
+
+@register('projects_by_scm_type')
+def projects_by_scm_type(since):
+    counts = dict(
+        (t[0] or 'manual', 0)
+        for t in models.Project.SCM_TYPE_CHOICES
+    )
+    for result in models.Project.objects.values('scm_type').annotate(
+        count=Count('scm_type')
+    ).order_by('scm_type'):
+        counts[result['scm_type'] or 'manual'] = result['count']
+    return counts
+
+
+@register('job_counts')   # TODO: evaluate if we want this (was not an ask)  Also, may think about annotating rather than grabbing objects for efficiency 
+def job_counts(since):    # TODO: Optimize -- for example, all of these are going to need to be restrained to the last 24 hours/INSIGHTS_SCHEDULE
+    counts = {}
+    counts['total_jobs'] = models.UnifiedJob.objects.all().count()
+    counts['successful_jobs'] = models.UnifiedJob.objects.filter(status='successful').count()
+    counts['cancelled_jobs'] = models.UnifiedJob.objects.filter(status='canceled').count()
+    counts['failed_jobs'] = models.UnifiedJob.objects.filter(status='failed').count()               
+    counts['error_jobs'] = models.UnifiedJob.objects.filter(status='error').count()   # Do we also want to include error, new, pending, waiting, running?
+    counts['new_jobs'] = models.UnifiedJob.objects.filter(status='new').count()       # Also, how much of this do we want `per instance`
+    counts['pending_jobs'] = models.UnifiedJob.objects.filter(status='pending').count()
+    counts['waiting_jobs'] = models.UnifiedJob.objects.filter(status='waiting').count()
+    counts['running_jobs'] = models.UnifiedJob.objects.filter(status='running').count()
+    for instance in models.Instance.objects.all():
```

Comment:
This is a _lot_ of nested queries.  Let's find a way to condense this into one aggregate query, like:

```sql
SELECT main_instance.uuid, COUNT(main_unifiedjob.id), status FROM main_unifiedjob, main_instance WHERE main_unifiedjob.execution_node=main_instance.hostname GROUP BY main_instance.id, main_unifiedjob.status;
                 uuid                 | count |   status
--------------------------------------+-------+------------
 ec2c225e-c00d-5ef8-71c6-0bdd2490d6d2 |     6 | canceled
 ec2c225e-c00d-5ef8-71c6-0bdd2490d6d2 |     1 | error
 ec2c225e-c00d-5ef8-71c6-0bdd2490d6d2 |   326 | failed
 ec2c225e-c00d-5ef8-71c6-0bdd2490d6d2 |  2796 | successful
(4 rows)
```

Reply:
From this data, we could do some math to _calculate_ `Instance.jobs_running` (it's the total of the count column, grouped by uuid), and `Instance.jobs_running` (it's the value of the row with `status='running'` or `0` if there is no row).